### ISSUE TITLE
Make pyimagej more easily usable without conda

### ISSUE THREAD

**Issue by hinerm** [imagej/pyimagej#88]
If pyimagej doesn't install cleanly across all platforms via `pip`, it may be a blocker to moving forward with a Cell Profiler bridge.

**Comment by hinerm** [comment]
@elevans @ctrueden fyi

**Comment by 0x00b1** [comment]
Yes, it would need to install cleanly (i.e. `pip install pyimagej`) on Linux, macOS, and Windows.

**Comment by ctrueden** [comment]
Yes, we need to improve this! And the good news is: it should be straightforward to do so as part of our switch from PyJNIus to JPype. 😄 

On the Python side, PyImageJ already installs cleanly. That is: I believe all Python dependencies are declared appropriately. The current issues are as follows:

1. The pyjnius library needs a `pyjnius.jar` file available locally in order to start up the JVM properly. An environment variable `PYJNIUS_JAR` currently needs to be set pointing to this JAR file. Before the JPype switch, our plan had been to enhance scyjava to lean on jgo to download this JAR from the remote Maven repository, same as all the other JARs, obviating the need to have it available in advance. But we hadn't done it yet. This is/was the most major reason that using conda was preferred: because unlike with pip, installing pyjnius via conda also installed this JAR file. But with the switch to JPype, I believe this issue will be moot.

2. The conda package also depends on openjdk and maven, neither of which are Python packages. With conda this is possible, but with pip it's less easy. My understanding is that we _could_ potentially bundle up openjdk and/or Maven as pip packages, but I rather heavily lean away from doing this, because: A) I don't want to be stuck maintaining it; and B) I don't think it will be necessary. It should be fine to instead make users responsible for: A) setting `JAVA_HOME` to point to a valid Java installation; and B) ensure `mvn` is on the system path. For CellProfiler specifically, (B) won't even be necessary because the new plan is to simply point CP at a local Fiji installation, and wrap it using pyimagej, using the Java that comes bundled with Fiji. So users won't need to install Java nor Maven. :smile:

See also [this forum thread](https://forum.image.sc/t/how-do-i-install-pyimagej/23189/4) with instructions on how to get the current version of pyimagej working without conda.

**Comment by ctrueden** [comment]
The ImageJ and CellProfiler teams met today to discuss the status of ImageJ integration into CellProfiler via pyimagej. The main concern, once the JPype migration is complete, will be how exactly CellProfiler should deal with the Java and potential Maven dependencies. Here's a summary of potential ways forward we discussed:

1. Instruct users to install Fiji separately, and tell CellProfiler the path to their installation via a configuration dialog. Then use that Fiji installation's bundled JVM and bundled JAR files.
- Pros:
 -- Would give CP access to the full power of Fiji, fully customizable/extensible by the user.
 -- Would simplify the CellProfiler build system and slim down its distribution (no more need to ship Java with CP).
- Cons:
 -- Users would need to perform an extra step to gain access to Bio-Formats functionality.
 -- Would need to update CP's other usage of Java (some kind of metadata parsing, IIUC) to work differently:
  --- Could reuse Fiji's JVM for that too, but then every CP user would need to install Fiji :-(
  --- Could rewrite the metadata parsing module in Python, but unclear how feasible that would be.

2. As 1, but keep shipping a JVM with CP and use that for CP's Java-related usages.
- Pros:
 -- Would let CP's other existing usage of Java keep working out of the box, even if the user does not configure Fiji.
- Cons:
 -- Would need to keep shipping a JVM with CP.
 -- CP's JVM (Java 14 right now) may not work as well with Fiji. Right now, we haven't tested ImageJ much with Java 14, and it's not an LTS release of Java.

3. Ship a Fiji installation with CP.
- Pros:
 -- All CP and CP+IJ features would work out of the box with no configuration.
- Same cons as (2), plus:
 -- Would bloat the CP distro by hundreds of MB.

4. As (3), but ship only the needed JAR files with CP.
- Pros:
 -- All desired CP features would work out of the box with no configuration.
 -- Smaller footprint than (3)—on the order of dozens of MB of JAR files, depending which features we want to ship.
 -- For additional user-desired features, they could still let users point at a customized ImageJ/Fiji installation.
- Same cons as (2), plus:
 -- CP's build system would need to invoke tooling to pull down the wanted JARs + dependencies into the distribution. Maven would be the obvious choice. IMO (without knowledge of the CP build system), this should be an easy thing to do.

5. As (4), but with JAR files published to PyPI, so that CP can pull down the JARs as Python modules.
- Pros:
 -- Would make it straightforward to add imagej to CP as a dependency, without modifying the CP build system.
- Cons:
 -- Would require additional packaging effort on the ImageJ side, wrapping hundreds of JAR files into Python modules, and publishing them to PyPI with the same dependency tree.
 -- Or, we could publish ImageJ to PyPI as a single uber-JAR (would be ~150MB). But then the benefits of ImageJ2's modularity would be lost on the Python side.
 -- It seems redundant to deploy JARs to PyPI when Maven repositories already offer binary deployment of Java artifacts, and jgo enables direct consumption of these artifacts (see next option) in a more flexible way.

6. Use pyimagej's support for remote endpoints (via jgo/Maven).
- Pros:
 -- Transparently use ImageJ and other Java features from CP both on the desktop and on the cluster, without different case logic.
 -- Avoids upfront shipping of JAR files with CP, in favor of downloading needed Java libraries on demand.
 -- Decouples the Java-side component versions from the Python-side component versions, so users can e.g. easily switch between different versions of ImageJ and Fiji components within the same CP installation.
- Cons:
 -- Would require either: A) shipping a copy of Maven with CP; or B) asking users to install Maven manually.
 -- Not as easy for users to customize their Fiji installations to gain access to additional functionality from CP. But you could combine this approach with the option for users to specify a locally installed Fiji as well, as discussed above, to get the best of both worlds.
 -- Substantial bootstrapping period the first time Java-side functionality is invoked, while jgo/Maven downloads and cache artifacts. Would be good to show the user a progress bar during initial bootstrapping.

7. As (6), but enhance jgo to have a "Maven-free" (i.e. pure Python) mode of operation. For our purposes, this seems doable to me, although it would be some development effort.
- Pros same as (6), plus:
 -- No need to ship Maven with CP, or ask users to install Maven, or bootstrap Maven ourselves.
- Cons:
 -- Extra dev effort to reimplement needed dependency reasoning on the Python side.

The CP team will discuss the above options, and form a consensus on how they'd like to proceed.

**Comment by Thrameos** [comment]
Not sure if this is relevant but for one of our internal projects we create mini distributions of Java using jlink .  It isn't exactly small (75M).

```
import os
​
# get the list of dependent modules 
modules = os.popen("jdeps --list-deps *.jar").read()
modules =  modules.strip().replace('\n',',').replace(" ","")
print("Dependent modules: " + modules)
# build the jre
print("Building custom JRE")
options = "--no-header-files --no-man-pages"
output = "jre"
os.system("jlink  " + options + " --add-modules " + modules + " --output " + output)
```

Another option is to pull a minimal Java distribution either using pip or some other tool and then use Apache Ivy to pull in jars.   It is much lighter than Maven if you are just trying to collect together a set of artifacts.

**Comment by ctrueden** [comment]
pyimagej 1.0.0 is now released on PyPI. Except for the non-Python dependencies of OpenJDK and Maven, it now cleanly installs and works with pip, thanks in large part to JPype's awesomeness.
